File: data_structures/python/BST.py
class Node:
    def __init__(self, data):
        self.left = None
        self.right = None
        self.data = data

class BST:
    def __init__(self):
        self.root = None

    def insert(self, root, node):
        if self.root is None:
            self.root = node

        else:
            if root.data < node.data:
                if root.right is None:
                    root.right = node
                else:
                    return self.insert(root.right, node)
            else:
                if root.left is None:
                    root.left = node
                else:
                    return self.insert(root.left, node)


    def search(self, root, data):
        if root is None:
            return False

        elif root.data == data:
            return True

        elif root.data < data:
            return self.search(root.right, data)

        else:
            return self.search(root.left, data)

    def inorder(self, root):
        if root:
            self.inorder(root.left)
            print(root.data)
            self.inorder(root.right)

    def preorder(self, root):
        if root:
            print(root.data)
            self.preorder(root.left)
            self.preorder(root.right)

    def postorder(self, root):
        if root:
            self.preorder(root.left)
            self.preorder(root.right)
            print(root.data)

my_bst = BST()
my_bst.insert(my_bst.root, Node(5))
my_bst.insert(my_bst.root, Node(1))
my_bst.insert(my_bst.root, Node(7))
my_bst.inorder(my_bst.root)

print(my_bst.search(my_bst.root, 7))

File: data_structures/java/BST.java
public class BST<T extends Comparable<T>> {
    public TreeNode<T> root;

    public BST() {
        this.root = null;
    }

    public void insert(TreeNode<T> root, T value) {
        if (this.root == null) {
            this.root = new TreeNode<T>(value);
        } else {
            if (root.data.compareTo(value) < 0) {
                if (root.right == null) {
                    root.right = new TreeNode<T>(value);
                } else {
                    insert(root.right, value);
                }
            } else {
                if (root.left == null) {
                    root.left = new TreeNode<T>(value);
                } else {
                    insert(root.right, value);
                }
            }
        }
    }

    public boolean search(TreeNode<T> root, T value) {
        if (root == null) {
            return false;
        } else if (root.data == value) {
            return true;
        } else if (root.data.compareTo(value) < 0) {
            return search(root.right, value);
        } else if (root.data.compareTo(value) > 0) {
            return search(root.left, value);
        }
        return false;
    }

    public void inorder(TreeNode<T> root) {
        if (root != null) {
            inorder(root.left);
            System.out.println(root.data);
            inorder(root.right);
        }
    }

    public void preorder(TreeNode<T> root) {
        if (root != null) {
            System.out.println(root.data);
            inorder(root.left);
            inorder(root.right);
        }
    }

    public void postorder(TreeNode<T> root) {
        if (root != null) {
            inorder(root.left);
            inorder(root.right);
            System.out.println(root.data);
        }
    }

    public static void main(String[] args) {
        BST<Integer> my_bst = new BST();
        my_bst.insert(my_bst.root, 5);
        my_bst.insert(my_bst.root, 1);
        my_bst.insert(my_bst.root, 7);
        my_bst.inorder(my_bst.root);
        System.out.println(my_bst.search(my_bst.root, 7));
    }
}
